teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,612 @@
+
+### Goals
+<!--
+List the specific goals of the TEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+**Simplicity**
+
+`Conditions` actually manifest themselves as `Tasks` but are implemented as a separate CRD which makes them complex. Moreover, maintaining the separate Condition CRD takes extra and unnecessary effort, given that it's really a `Task` underneath. We prefer to reuse existing components when possible - in this case, reusing `Tasks` as `Conditions` - to make guarded execution of `Tasks` in Tekton simple. 
+
+**Efficiency**
+
+Checking `Conditions` is slow and expensive because it spins up new `Pods` to evaluate each `Condition`. For example, Tekton dogfooding has been heavily using `Conditions` to decide what to run e.g only run this branch if it contains a specific type of file. These small checks add up such that many pods are used to check `Conditions` and it becomes slow and expensive. Even worse, we don't have a conditional branching construct (if/else or switch), so users have to implement and execute opposite `Conditions` which makes it even slower and more expensive. This can also be a problem in terms of the resource limits, requests, quotas and LimitRanges. 
+
+**Skipping**
+
+When a `Condition` fails, the guarded `Task` and its branch (dependent `Tasks`) are skipped. A `Task` is dependent on and in the branch of another `Task` as specified by ordering using `runAfter` or by resources using `Results`, `Workspaces` and `Resources`.  In some use cases of `Conditions`, when a `Condition` evaluates to `False`, users need to skip the guarded `Task` only and allow dependent `Tasks` to execute. An example use case is when there’s a particular `Task` that a Pipeline wants to execute when the git branch is dev/staging/qa, but not when it’s the main/master/prod branch. Another use case is when a user wants to send out a notification whether or not a parent guarded `Task` was executed, as described in [this issue](https://github.com/tektoncd/pipeline/issues/2937). 
+
+### Non-Goals
+<!--
+What is out of scope for this TEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+**Status**
+    
+It is currently difficult to distinguish between a `Task` that was skipped due to `Condition` evaluating as `False` and a `Task` that failed. This is because we reuse `ConditionSucceeded` from Knative which can have the following states: `True`, `False` or `Unknown`. When a `Task` either fails or is skipped from `Condition` evaluating to `False`, we mark `ConditionSucceeded` as `False`. 
+
+## Requirements
+<!--
+List the requirements for this TEP.
+-->
+1. Users need a way to specify complex `Conditions`, such as those needing access to files in a `Workspace`. 
+1. Users need a way to specify simple `Conditions`, such as checking the name of a git branch matches.
+1. Users need a way to specify what happens when `Conditions` evaluate to `False`.
```

Comment:
might want to elaborate on this one a bit more, "what happens when" kinda sounds to me like it could cover many different things (e.g. execute Task A when false, Task B when true)